### PR TITLE
Iso8601-compliant dateTime parsing

### PR DESCRIPTION
**Pull request by JensMertelmeyer**
This is a suggestion for issue #192: When running tests from people with other date formats, the tests will pass on one machine, and fail on the other. Example: `[TestCase('today','19/07/2017 01:00:00')]` is a valid representation for some countries, but will fail on a system with the `EN-US` locale.

I have expanded the TValue → TDateTime/TDate/TTime conversion to try again with an ISO8601 representation if the conversion using the current locale has failed. This way, `[TestCase('today','2017-07-19 01:00:00')]` (and others) should work on all test systems, independent of the current locale.

Existing behaviour is not altered, it will still pass on all date/time/dateTime arguments that are not in an ISO8601 representation as long as the current locale matches.

I haven't yet worked with DUnitX, so I am not sure if the test cases are sufficient. I have been unable to write test cases that set the current locale to something like `DE-DE`, `IT` or `EN-US` and then try to pass parameters into a function by using the `TestCase` attribute.

Milliseconds are ignored.

### CODE REVIEW

**Comment by vincentparrett** [comment]
Thanks, this looks good. You should be able to test different locale's in a hacky way by 

1) take a copy of the current settings for System.SysUtils.FormatSettings 
2) modify FormatSettings to the locale you want to test
3) run the test
4) restore the FormatSettings to the original value.

I have not tried this, just a thought.

**Comment by JensMertelmeyer** [comment]
Yes, that's what I thought. However, I have been unable to write DUnitX tests for this scenario.

My plan was to write tests that modify the format settings in a `[SetupFixture]` and restore in a `[TearDownFixture]` method while doing the locale-specific tests in between. I think this is not possible because

1. TDUnitXTestRunner.Execute() calls BuildFixtures()
2. This will parse all the attributes and generate test cases
3. TDUnitXTestRunner.Execute() calls ExecuteFixtures(..)
4. Which will execute the setup fixtures, then tests

The core problem is that DUnitX will first parse all the attributes to generate test cases, and then give my code a chance to run ☹

Writing tests for the `TValue` locale-specific conversion + fallback in `Source/DUnitX.Utils.pas` however, is trivial. Do you think this would be sufficient?

File: Source/DUnitX.Utils.pas [hunk 1]
```diff
@@ -1121,6 +1121,58 @@ function ConvStr2Enum(const ASource: TValue; ATarget: PTypeInfo; out AResult: TV
   Result := True;
 end;
 
+function Str2DateTime(const ASource: String): TValue;
```
Comment (vincentparrett) [inline Source/DUnitX.Utils.pas:1124]: Rename to Str2DateTimeValue to be more explicit about what it does

Comment (vincentparrett) [reply Source/DUnitX.Utils.pas:1124]: Same for the other functions below.

File: Tests/DUnitX.Tests.Example.pas [hunk 2]
```diff
@@ -141,7 +158,7 @@   TExampleFixture3 = class
 implementation
 
 uses
-  DUnitX.DUnitCompatibility;
+  DUnitX.DUnitCompatibility, System.DateUtils;
```
Comment (vincentparrett) [inline Tests/DUnitX.Tests.Example.pas:161]: Needs  {$IFDEF USE_NS} like the other units to ensure compatibility with D2010/XE.

File: Source/DUnitX.Utils.pas [hunk 3]
```diff
@@ -1121,6 +1121,58 @@ function ConvStr2Enum(const ASource: TValue; ATarget: PTypeInfo; out AResult: TV
...
+var
+  dateTime: TDateTime;
+  iso8601: TFormatSettings;
+begin
+  if(not TryStrToDateTime(ASource, dateTime)) then
+    begin
+      iso8601 := TFormatSettings.Invariant();
```
Comment (vincentparrett) [inline Source/DUnitX.Utils.pas:1131]: This fails in XE7 - which version was Invariant added in - need to find a way to get this that works from D2010

Comment (vincentparrett) [reply Source/DUnitX.Utils.pas:1131]: Invariant was added in XE8 - I have pushed a change in master that adds a record helper for TFormatSettings in DUnitX.Utils that should resolve this, please update your pr branch from master and push again.

**Comment by JensMertelmeyer** [comment]
I have added 26 test cases for the conversion String -> TValue -> TDateTime/TDate/TTime

One should be aware that the tests can fail if the machine is using one of the used locales (EN-GB, EN-US or DE) and has also overridden the format settings in Windows. It can be circumvented, but would require adding more type helpers to TFormatSettings (which didn't even have a constructor back in 2010 or XE).

**Comment by vincentparrett** [comment]
Nice work @JensMertelmeyer  - thanks.

**Pull request merged**